From a00e71c75f3aa5271970f7e60aea71c962f1d380 Mon Sep 17 00:00:00 2001
From: Azure Sphere Team <azuresphereoss@microsoft.com>
Date: Thu, 24 Aug 2017 00:18:26 +0000
Subject: [PATCH] Support for MT3620

Add new dap mem-ap command needed to toggle coresight GPR registers prior to target examine.

---
diff --git a/src/target/arm_adi_v5.c b/src/target/arm_adi_v5.c
index eafc2ddc..52b3fee1 100644
--- a/src/target/arm_adi_v5.c
+++ b/src/target/arm_adi_v5.c
@@ -1811,6 +1811,47 @@ COMMAND_HANDLER(dap_apid_command)
 	return retval;
 }
 
+COMMAND_HANDLER(dap_memap_command)
+{
+	struct adiv5_dap *dap = adiv5_get_dap(CMD_DATA);
+
+	uint32_t apsel, reg_addr, value;
+	int retval;
+
+	//ap_num reg_addr [value]
+	if (CMD_ARGC != 2 && CMD_ARGC != 3)
+		return ERROR_COMMAND_SYNTAX_ERROR;
+	
+	COMMAND_PARSE_NUMBER(u32, CMD_ARGV[0], apsel);
+	/* AP address is in bits 31:24 of DP_SELECT */
+	if (apsel >= 256)
+		return ERROR_COMMAND_SYNTAX_ERROR;
+
+	COMMAND_PARSE_NUMBER(u32, CMD_ARGV[1], reg_addr);
+
+	// We need to init first
+	retval = dap_dp_init(dap);
+	if (retval != ERROR_OK) {
+		LOG_ERROR("Could not initialize the debug port");
+		return retval;
+	}
+
+	if (CMD_ARGC == 3) {
+		COMMAND_PARSE_NUMBER(u32, CMD_ARGV[2], value);
+
+		command_print(CMD, "Setting MEM-AP register @ 0x%8.8" PRIx32 " to 0x%8.8" PRIx32, reg_addr, value);
+		
+		retval = mem_ap_write_atomic_u32(dap_ap(dap, apsel), reg_addr, value);
+	} else {
+		retval = mem_ap_read_atomic_u32(dap_ap(dap, apsel), reg_addr, &value);
+		if (retval == ERROR_OK) {
+			command_print(CMD, "MEM-AP register @ 0x%8.8" PRIx32 " value is 0x%8.8" PRIx32, reg_addr, value);
+		}
+	}
+
+	return retval;
+}
+
 COMMAND_HANDLER(dap_apreg_command)
 {
 	struct adiv5_dap *dap = adiv5_get_dap(CMD_DATA);
@@ -2025,6 +2025,13 @@ static const struct command_registration dap_commands[] = {
 			"bus access [0-255]",
 		.usage = "[cycles]",
 	},
+	{
+		.name = "memap",
+		.handler = dap_memap_command,
+		.mode = COMMAND_EXEC,
+		.help = "Set / read MEM-AP register value (supports modification prior to target examination)",
+		.usage = "ap_num reg_addr [value]",
+	},
 	{
 		.name = "ti_be_32_quirks",
 		.handler = dap_ti_be_32_quirks_command,
-- 
2.17.1.windows.2

